### PR TITLE
feat(core): support func type when change builderConfig in plugin

### PR DESCRIPTION
**Pull request by 10Derozan**
## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/shared/src/types/Plugin.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export interface RspressPlugin {
   /**
    * Rsbuild config.
    */
-  builderConfig?: RsbuildConfig;
+  builderConfig?: RsbuildConfig | ((config: RsbuildConfig, isSSR: boolean) => void | RsbuildConfig);
```
Comment (chenjiahan) [inline packages/shared/src/types/Plugin.ts:36]: the second param should be an object, making it  easier to add more params in the future.

File: packages/core/src/node/initRsbuild.ts [hunk 2]
```diff
@@ -239,15 +239,13 @@ export async function initRsbuild(
     pluginDriver,
   );
 
+  const builderConfigList = [...pluginDriver.getPlugins().map(plugin => plugin.builderConfig ?? {}), config?.builderConfig || {}, extraRsbuildConfig || {}]
+  const rsbuildConfig = builderConfigList.reduce<RsbuildConfig>((acc, cur): RsbuildConfig => {
+    return cur instanceof Function ? cur(acc, isSSR) || acc : mergeRsbuildConfig(acc, cur);
+  }, internalRsbuildConfig)
+
```
Comment (chenjiahan) [inline packages/core/src/node/initRsbuild.ts:246]: can we add a test case for this change

Comment (chenjiahan) [reply packages/core/src/node/initRsbuild.ts:246]: also document

File: packages/shared/src/types/Plugin.ts (hunk 1 above)
Comment (chenjiahan) [reply packages/shared/src/types/Plugin.ts:36]: The "config" hook supports function type, maybe we should use it to modify builderConfig.

Comment (10Derozan) [reply packages/shared/src/types/Plugin.ts:36]: The config hook don't have internalRsbuildConfig, derived from the rspress configuration.
The config hook still process rspress config,.

Comment (chenjiahan) [reply packages/shared/src/types/Plugin.ts:36]: Sure, if you need to process the builtin Rsbuild config, you can write an Rsbuild plugin.

**Pull request closed without merging**